src/DynamoCore/Graph/Nodes/CodeBlockNode.cs

```diff
@@ -239,6 +239,73 @@ public string Code
             private set { code = value; }
         }
 
+        /// <summary>
+        ///     Contains data needed to reconstruct input and output ports during deserialization 
+        ///     when there is an error in a code block node.
+        /// </summary>
+        public class ErrorStatePortInfo
+        {
+            internal List<string> inPortNames;
+            internal List<int> outPortLineIndexes;
+
+            internal ErrorStatePortInfo(
+                IEnumerable<PortModel> inPorts,
+                IEnumerable<PortModel> outPorts)
+            {
+                inPortNames = new List<string>();
+                foreach (var inPort in inPorts)
+                    inPortNames.Add(inPort.Name);
+
+                outPortLineIndexes = new List<int>();
+                foreach (var outPort in outPorts)
+                    outPortLineIndexes.Add(outPort.LineIndex);
+            }
+
+            /// <summary>
+            ///     Contains data needed to reconstruct input ports during deserialization 
+            ///     when there is an error in a code block node.
+            /// </summary>
+            public List<string> InPortNames
+            {
+                get
+                {
+                  return inPortNames;
+                }
+            }
+
+            /// <summary>
+            ///     Contains data needed to reconstruct output ports during deserialization 
+            ///     when there is an error in a code block node.
+            /// </summary>
+            public List<int> OutPortLineIndexes
+            {
+                get
+                {
+                  return outPortLineIndexes;
+                }
+            }
+        }
+
+        /// <summary>
+        ///     Contains data needed to reconstruct input and output ports during deserialization 
+        ///     when there is an error in a code block node.
+        ///     Note that if the code block node is not in an error state the data returned will be null.
+        ///     This is to avoid serializing the error data if it is not needed.
+        ///     Also note that this property is only for JSON serialization, and is here to avoid needing
+        ///     to create a special converter for code block nodes (due to the complicated serialization
+        ///     of code block nodes).
+        /// </summary>
+        public ErrorStatePortInfo ErrorStatePortData
```

Comment:
this will generate a serialized version of a codeblock which does not adhere to the schema... I guess I am not clear why we need to serialize this extra information and modify the schema at this late stage.

Couldn't we just deserialize the port information and pass those ports to the `codeBlockNodeModel` constructor, and if the node is in an error state not remove the ports we just deserialized? This would not require serializing any other information.

Reply:
Maybe I don't understand what's going on, but this comment seems rather odd.  Why would you *ever* remove the ports that are deserialized from the JSON?

Reply:
code blocks regenerate and remove their ports all the time, for example whenever they deserialize.

I think this task can be accomplished by using the already existing port information in the schema, not requiring a new class and property on codeblocks. (I could be wrong) but I would like @ColinDayOrg to look at that before we add new stuff to the schema for this.

Reply:
That made sense with the XML format, as the ports weren't stored in the format.  I don't think that makes sense with the JSON format, since it explicitly stores the ports.